src/main/java/net/schmizz/sshj/common/CircularBuffer.java

```diff
@@ -0,0 +1,151 @@
+/*
+ * Copyright (C)2009 - SSHJ Contributors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package net.schmizz.sshj.common;
+
+public class CircularBuffer<T extends CircularBuffer<T>> {
+
+    public static class CircularBufferException
+            extends SSHException {
+
+        public CircularBufferException(String message) {
+            super(message);
+        }
+    }
+
+    public static final class PlainCircularBuffer
+            extends CircularBuffer<PlainCircularBuffer> {
+
+        public PlainCircularBuffer(int size, int maxSize) {
+            super(size, maxSize);
+        }
+    }
+
+    private int getNextSize(int i) {
+        // Use next power of 2.
+        int j = 1;
+        while (j < i) {
+            j <<= 1;
+            if (j <= 0) throw new IllegalArgumentException("Cannot get next power of 2; "+i+" is too large");
+        }
+        return Math.min(j, maxSize); // limit to max size
+    }
+
+    private final int maxSize;
+    private byte[] data;
+    private int rpos;
+    private int wpos;
+
+    public CircularBuffer(int size, int maxSize) {
+        this.maxSize = maxSize;
+        int initialSize = getNextSize(size);
+        if (initialSize > maxSize) {
+            throw new IllegalArgumentException(
+                String.format("Initial power of 2 size %d larger than maximum size %d", initialSize, maxSize));
+        }
+        this.data = new byte[initialSize];
+        this.rpos = 0;
+        this.wpos = 0;
+    }
+
+    public int available() {
+        int available = wpos - rpos;
+        return available >= 0 ? available : available + data.length; // adjust if wpos is left of rpos
```

Comment:
It this right? If I have a buffer that's of size 4, and `rsize == 2` while `wsize == 1` (`data.length` would have to be 3, as all bytes are unread between offset 2, wrapping back up to 1, so bytes 2, 3, and 1 are used up). In this case, `available` should be 1 (there's only 1 byte left to write to after `wpos` until it reaches up to `rpos`). With the math above, you would return 2 (`-1 + data.length == 2`).

Should this not return `-available` in the case where `wpos` is to the left of `rpos`, or do I need another cup of coffee?